### PR TITLE
Feature/consolidate proto

### LINKED ISSUE #47: `APPEND` Optional topic parameter
Since topics always begin with a leading /, how bad of an idea would it be to assume that if data doesn't look like it begins with a topic, we default to "/"?

--

I had thought about doing optional topics in the PR, but decided to hold off. I think I want to add the length based message shape to the AppendRequest as well to make it very easy to do. This will require the REPL client to parse the plaintext into the proper message, but that change will be better for an API.

### LINKED ISSUE #50: Consolidate protocol pipeline functions
Unify sending, receiving, and parsing functions on a format.

- Message Parsing
- Fixed width command
- Perf

### PR DESCRIPTION
**Pull request by gideonw**
Closes #47 
Closes #50

### CODE REVIEW

File: cmd/fossil/client/client.go [hunk 1]
```diff
@@ -155,24 +176,25 @@ func clientPrompt(c net.Conn) {
...
+		msgb, err := replMsg.Marshal()
 		if err != nil {
-			fmt.Printf("Err: unable to send command\n\t'%s'\n", err)
+			log.Error().Err(err).Msg("error marshaling message into bytes")
+			continue
 		}
-
-		resp, err := proto.ReadBytes(c)
+		_, err = c.Write(msgb)
+		// err = send(c, line)
```
Comment (dburkart) [inline cmd/fossil/client/client.go:187]: Remove commented code :)

File: docs/protocol.md [hunk 2]
```diff
@@ -46,6 +50,11 @@ TODO
 #### AppendRequest
 ```
 topic data
++--------+----------------+--------------+
+|   4    |       N        |  4 + N -> M  |
++--------+----------------+--------------+
+|  len   |     topic      |   Data...    |
++--------+----------------+--------------+
```
Comment (dburkart) [inline docs/protocol.md:57]: APPEND data section has it's own length prefix independent of the overall message?

File: pkg/proto/message.go [hunk 3]
```diff
@@ -27,23 +27,44 @@ var (
...
-	b := make([]byte, length)
-	_, err = io.ReadFull(r, b)
+// func ReadMessage(r io.Reader) (Message, error) {
+// 	data, err := ReadBytes(r)
+// 	if err != nil {
+// 		return Message{}, fmt.Errorf("error reading message")
+// 	}
+
+// 	return ParseMessage(data)
+// }
```
Comment (dburkart) [inline pkg/proto/message.go:42]: Remove le commented code

File: pkg/proto/message.go [hunk 4]
```diff
@@ -27,23 +27,44 @@ var (
...
+// 		return nil, errors.New("unable to read length prefix")
+// 	}
+// 	length := binary.LittleEndian.Uint32(lengthPrefix)
+// 	b := make([]byte, length)
+// 	_, err = io.ReadFull(r, b)
+// 	if err != nil {
+// 		return nil, fmt.Errorf("unable to read response\n\t'%s'", string(b))
+// 	}
+// 	return b, nil
+// }
```
Comment (dburkart) [inline pkg/proto/message.go:66]: Here as well

File: pkg/proto/message.go [hunk 5]
```diff
@@ -69,27 +90,53 @@ func NewMessageWithType(cmd string, t Marshaler) Message {
...
+// 	if ind == -1 {
+// 		return ret, fmt.Errorf("malformed message")
+// 	}
+// 	ret.Command = strings.ToUpper(string(b[0:ind]))
+// 	if ind < len(b) {
+// 		ret.Data = b[ind+1:]
+// 	}
+
+// 	return ret, nil
+// }
```
Comment (dburkart) [inline pkg/proto/message.go:108]: Also this

**Review by dburkart — COMMENTED** [review]
Looking good so far. Just one question, and a few nits.

File: pkg/repl/parser.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	// Marshal message based on the command
+	switch strings.ToUpper(string(cmd)) {
+	case proto.CommandAppend:
+		req := proto.AppendRequest{}
+		data := b[ind+1:]
+
+		// check for topic
+		pathInd := bytes.IndexByte(data, '/')
+		// check for space after topic, no space means the data starts with /
+		spaceInd := bytes.IndexByte(data, ' ')
```
[the remark this answers is not in the document]
Comment (gideonw) [reply pkg/repl/parser.go:42]: Good call, I will also add a test for this.

File: docs/protocol.md (hunk 2 above)
Comment (dburkart) [reply docs/protocol.md:57]: Ahh, I see. But data could start with '/' so long as the client explicitly set the topic. For example:

`APPEND / /foo`

Should work

**Comment by gideonw** [comment]
@dburkart FYI, I followed up with my APPEND change and removed the space delim from the remainder of the message types since we don't need that for parsing since the len is a fixed width.

**Pull request merged**